File: programmers/kakao#2/java/KAKAO17.java
public class KAKAO17 {
    class Solution {
        public String solution(String m, String[] musicinfos) {
            String answer = "(None)";
            int maxPlay = 0;
            m = process(m);
            
            for (String musicinfo : musicinfos) {
                String[] infoArr = musicinfo.split(",");
                infoArr[3] = process(infoArr[3]);
                int start = Integer.parseInt(infoArr[0].split(":")[0]) * 60 + Integer.parseInt(infoArr[0].split(":")[1]);
                int end = Integer.parseInt(infoArr[1].split(":")[0]) * 60 + Integer.parseInt(infoArr[1].split(":")[1]);
                
                int diff = end - start;
                
                StringBuffer sb = new StringBuffer();
                for(int i=0;i<diff/infoArr[3].length();i++) {
                    sb.append(infoArr[3]);
                }
                sb.append(infoArr[3].substring(0, diff % infoArr[3].length()));
                
                String music = sb.toString();
    
                if (music.indexOf(m) != -1) {
                    if (maxPlay < diff) {
                        answer = infoArr[2];
                        maxPlay = diff;
                    }
                }
            }
            return answer;
        }
        public String process(String str){
            return str.replace("A#", "a").replace("C#", "c").replace("D#", "d").replace("F#", "f").replace("G#", "g");
        }
    }
}
/**
 * [카카오 2018 공채] 방금 그 곡
 * 해결: O
 */

File: 0_4.programmers/kakao#2/python/KAKAO17.py
def solution(m, musicinfos):
    m = process(m)
    max_play = 0
    answer = "(None)"
    for musicinfo in musicinfos:
        start, end, title, info = musicinfo.split(",")
        
        info = process(info)
        
        start_h, start_m = map(int, start.split(":"))
        end_h, end_m = map(int, end.split(":"))
        
        times = end_h * 60 + end_m - start_h*60 - start_m
        
        music = info * (times // len(info)) + info[:times % len(info)]
        
        if m in music:
            if max_play < times:
                answer = title
                max_play = times
    
    return answer

def process(str):
    return str.replace("C#", "c").replace("D#", "d").replace("F#", "f").replace("G#", "g").replace("A#", "a")

"""
[2018 카카오 공채] 방금 그 곡
해결: O
"""